src/pages/iou/IOUModal.js

```diff
@@ -247,35 +263,61 @@ class IOUModal extends Component {
         if (this.state.currentStepIndex >= this.steps.length - 1) {
             return;
         }
+
         this.setState(prevState => ({
             previousStepIndex: prevState.currentStepIndex,
             currentStepIndex: prevState.currentStepIndex + 1,
         }));
     }
 
     /**
-     * Update comment whenever user enters any new text
+     * Checks if user has a GOLD wallet then creates a paid IOU report on the fly
      *
-     * @param {String} comment
+     * @param {String} paymentMethodType
      */
-    updateComment(comment) {
-        this.setState({
+    sendMoney(paymentMethodType) {
+        const hasGoldWallet = this.props.userWallet.tierName && this.props.userWallet.tierName === CONST.WALLET.TIER_NAME.GOLD;
+
+        // If the user is trying to send money, then they need to upgrade to a GOLD wallet
+        if (!hasGoldWallet) {
+            Navigation.navigate(ROUTES.IOU_ENABLE_PAYMENTS);
+            return;
+        }
+
+        const amount = Math.round(this.state.amount * 100);
+        const currency = this.props.iou.selectedCurrencyCode;
+        const comment = this.state.comment;
+
+        const newIOUReportDetails = JSON.stringify({
+            amount,
+            currency,
+            requestorEmail: this.state.participants[0].login,
+            comment,
+            idempotencyKey: Str.guid(),
+        });
+
+        IOU.payIOUReport({
+            chatReportID: lodashGet(this.props, 'route.params.reportID', ''),
+            reportID: 0,
+            paymentMethodType,
+            amount,
+            currency,
+            requestorPayPalMeAddress: this.state.participants[0].payPalMeAddress,
+            requestorPhoneNumber: this.state.participants[0].phoneNumber,
             comment,
+            newIOUReportDetails,
+            shouldRedirectToChatReport: true,
         });
```

Comment:
Basically moved the minimal `sendMoney` action logic from `libs/action/IOU.js` from there to here. Reason being:

- We are really only supposed to call `sendMoney` from the `IOUModal`. I couldn't think of a case where we'd want to expose a `sendMoney` action elsewhere.
- One of the original reasons for having a `sendMoney` action was that we wanted the aftereffect of redirecting the user back to the dm of the user they're paying and it breaks app philosophy to call a  `.then()` on any actions inside views. The way I thought I'd accomplish this originally inside a `sendMoney` action would be to wrap `payIOUReport` in a promise there then call `Navigation.navigate` from a `.then`. It seemed simpler to me to just pass an additional param to `payIOUReport` to handle the aftereffect.
- By doing this too, we can also avoid repeating passing the same params through an additional function.

Reply:
Great summary of the thought process here. Thanks!

Happy to leave things as they are for now, but just wanted to give this one last exploration as problems like these will pop up in the future and how we handle them sets the precedent for others to follow.

>We are really only supposed to call sendMoney from IOUModal. I couldn't think of a case where we'd want to expose a sendMoney action elsewhere.

It might be reasonable to extract this into an "action" regardless of code re-use benefits. It will help separate the "action" logic from the view, keep things organized, and we get the benefit of replacing `shouldRedirectToChatReport` with a `.finally()` in an `IOU.sendMoney()` method.

>we wanted the aftereffect of redirecting the user back to the dm of the user they're paying and it breaks app philosophy to call a .then() on any actions inside views

Could we still follow this philosophy if we create a new action? Why does this rule exist?

>wrap payIOUReport in a promise there then call Navigation.navigate from a .then. 

What if we returned the promise here instead of wrapping it with a promise?

https://github.com/Expensify/App/blob/ca34f199fa68b45b744594f11cd0aea03e5a4712/src/libs/actions/IOU.js#L281

Calling `.then()` like so should not break any app philosophy

```js
function sendMoneyAndNavigateToReport(params) {
    payIOUReport(params)
        .finally(() => {
            Navigation.navigate(ROUTES.REPORT);        
        });
}
```

>It seemed simpler to me to just pass an additional param shouldRedirectToChatReport to payIOUReport to handle the redirect aftereffect.

An advantage of navigating from outside this method (instead of within it) is that we pass the responsibility to the caller of the action and reduce the overall responsibility of the action itself. When will `shouldRedirectToChatReport` be true? Only when we are sending money. Does it depend on anything else in `payIOUReport`? Only the promise - which could be returned.

Reply:
> It might be reasonable to extract this into an "action" regardless of code re-use benefits. It will help separate the "action" logic from the view, keep things organized, and we get the benefit of replacing shouldRedirectToChatReport with a .finally() in an IOU.sendMoney() method.

I agree with this.

> Could we still follow this philosophy if we create a new action? Why does this rule exist?

Yes we could, and I believe this rule exists so we don't introduce any side effects within views (views should be responsible for displaying data is my understanding).

> What if we returned the promise here instead of wrapping it with a promise?

I guess I mostly struggled with how to make this work nicely with `asyncOpenURL` since that method doesn't return a promise. Wouldn't we have to wrap `asyncOpenURL` in another promise anyways or can we simply return it and attach the `.finally` handler to it?

Reply:
Sorry bout that, didn't notice `asyncOpenURL` was not returning a promise - but can think of two solutions here:

- Always return the promise passed to `asyncOpenURL()` instead of void

```js
export default function asyncOpenURL(promise, url) {
    if (!url) {
        return promise;
    }

    return promise.then((params) => {
        Linking.openURL(typeof url === 'string' ? url : url(params));
    });
}
```

- Attach all the handlers to the promise and do asyncOpenURL then return it

```js
    payIOUPromise.then((response) => {...});
    asyncOpenURL(payIOUPromise, url);
    return payIOUPromise;
```